src/simplnx/Utilities/ThreadedAlgorithm.hpp

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include "simplnx/Filter/IFilter.hpp"
+#include "simplnx/simplnx_export.hpp"
+
+#include <chrono>
+#include <cstdint>
+
+namespace nx::core
+{
+
+class SIMPLNX_EXPORT ThreadedAlgorithm
+{
+
+public:
+  ThreadedAlgorithm(const IFilter::MessageHandler& mesgHandler, const std::atomic_bool& shouldCancel);
+  virtual ~ThreadedAlgorithm() noexcept;
+
+  ThreadedAlgorithm(const ThreadedAlgorithm&) = delete;
+  ThreadedAlgorithm(ThreadedAlgorithm&&) noexcept = delete;
+  ThreadedAlgorithm& operator=(const ThreadedAlgorithm&) = delete;
+  ThreadedAlgorithm& operator=(ThreadedAlgorithm&&) noexcept = delete;
+
+  const std::atomic_bool& getCancel() const;
+
+  void sendThreadSafeProgressMessage(size_t counter);
+
+  void sendThreadSafeInfoMessage(const std::string& message);
+
+  const IFilter::MessageHandler& getMessageHandler() const;
+
+  void setTotalElements(size_t totalElements);
```

Comment:
There's nothing wrong with this method in particular, but its visibility is concerning since it modifies the object without a lock. This should never be called in parallel, but in its current state it can be done.

To get around this we could make this an arg that must be passed to the ctor.